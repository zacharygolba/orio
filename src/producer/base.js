// @flow

export default class ProducerBase<T> implements Iterator<T> {
  // eslint-disable-next-line prettier/prettier
  /*:: @@iterator: () => Iterator<T>; */

  // $FlowIgnore
  get [Symbol.toStringTag](): string {
    return this.constructor.name
  }

  // $FlowIgnore
  [Symbol.iterator](): Iterator<T> {
    return this
  }

  next(): IteratorResult<T, void> {
    throw new Error('unimplemented')
  }
}
