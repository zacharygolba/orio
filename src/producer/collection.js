// @flow

import type { Producer } from './types'

export class MapProducer<K, V> implements Producer<*> {
  size: number
  source: Iterator<[K, V]>
  /*:: @@iterator: () => Iterator<[K, V]> */

  constructor(source: Map<K, V>) {
    this.size = source.size
    this.source = source.entries()
  }

  // $FlowFixMe
  [Symbol.iterator](): Iterator<[K, V]> {
    return this
  }

  next(): IteratorResult<[K, V], void> {
    return this.source.next()
  }

  sizeHint(): number {
    return this.size
  }
}

export class SetProducer<T> implements Producer<T> {
  size: number
  source: Iterator<T>
  /*:: @@iterator: () => Iterator<T> */

  constructor(source: Set<T>) {
    this.size = source.size
    this.source = source.values()
  }

  // $FlowFixMe
  [Symbol.iterator](): Iterator<T> {
    return this
  }

  next(): IteratorResult<T, void> {
    return this.source.next()
  }

  sizeHint(): number {
    return this.size
  }
}
