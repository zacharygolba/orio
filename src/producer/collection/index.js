// @flow

export { default as IndexedProducer } from './indexed'
export { default as MapProducer } from './map'
export { default as SetProducer } from './set'
