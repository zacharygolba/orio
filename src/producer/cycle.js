// @flow

import type { IndexedCollection, Producer } from './types'

export default class CycleProducer<T> implements Producer<T> {
  source: IndexedCollection<T>
  state: number
  /*:: @@iterator: () => Iterator<T> */

  constructor(source: IndexedCollection<T>) {
    this.source = source
    this.state = 0
  }

  // $FlowFixMe
  [Symbol.iterator](): Iterator<T> {
    return this
  }

  next(): IteratorResult<T, void> {
    if (this.state >= this.source.length) {
      this.state = 0
    }

    return {
      done: false,
      value: this.source[this.state++],
    }
  }

  sizeHint(): number {
    return Infinity
  }
}
