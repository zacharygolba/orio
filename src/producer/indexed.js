// @flow

import * as result from '../result'

import type { IndexedCollection, Producer } from './types'

export default class IndexedProducer<T> implements Producer<T> {
  source: IndexedCollection<T>
  size: number
  state: number
  /*:: @@iterator: () => Iterator<T> */

  constructor(source: IndexedCollection<T>) {
    this.source = source
    this.size = source.length
    this.state = 0
  }

  // $FlowFixMe
  [Symbol.iterator](): Iterator<T> {
    return this
  }

  next(): IteratorResult<T, void> {
    if (this.state >= this.size) {
      return result.done()
    }

    return result.next(this.source[this.state++])
  }

  sizeHint(): number {
    return this.size
  }
}
