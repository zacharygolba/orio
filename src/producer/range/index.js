// @flow

export { default as CharProducer } from './char'
export { default as NumberProducer } from './number'
