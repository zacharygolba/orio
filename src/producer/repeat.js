// @flow

import * as result from '../result'

import ProducerBase from './base'

export default class RepeatProducer<T> extends ProducerBase<T> {
  value: T

  constructor(value: T) {
    super()
    this.value = value
  }

  next(): IteratorResult<T, void> {
    return result.next(this.value)
  }

  sizeHint(): number {
    return Infinity
  }
}
