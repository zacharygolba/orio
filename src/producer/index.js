// @flow

import Iter from '../iter'
// import type { IntoIterator } from '../types'

import { CharProducer, NumberProducer } from './range'
import { MapProducer, SetProducer } from './collection'
import CycleProducer from './cycle'
import IndexedProducer from './indexed'
import RepeatProducer from './repeat'
import UnboundProducer from './unbound'
import type { Producer } from './types'

export type { IndexedCollection, Producer } from './types'
export { MapProducer, SetProducer } from './collection'
export { default as CycleProducer } from './cycle'
export { default as IndexedProducer } from './indexed'
export { CharProducer, NumberProducer } from './range'
export { default as RepeatProducer } from './repeat'
export { default as UnboundProducer } from './unbound'

/*:: declare function createProducer<T>(any): Producer<T> */
export function createProducer<T>(source: any): Producer<T> {
  if (source == null) {
    return new IndexedProducer([])
  }

  if (Array.isArray(source) || typeof source === 'string') {
    return new IndexedProducer(source)
  }

  if (
    source instanceof CharProducer
    || source instanceof NumberProducer
    || source instanceof MapProducer
    || source instanceof SetProducer
    || source instanceof CycleProducer
    || source instanceof IndexedProducer
    || source instanceof RepeatProducer
    || source instanceof UnboundProducer
  ) {
    return source
  }

  if (source instanceof Iter) {
    return source.producer
  }

  if (source instanceof Set) {
    return new SetProducer(source)
  }

  if (source instanceof Map) {
    return new MapProducer(source)
  }

  if (typeof source[Symbol.iterator] === 'function') {
    return new UnboundProducer(source)
  }

  return new IndexedProducer([source])
}
