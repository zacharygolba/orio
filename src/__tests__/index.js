// @flow

import { CharProducer, CycleProducer, NumberProducer } from '../producer'
import Iter from '../iter'
import * as iter from '../'

test('.chars()', () => {
  expect(iter.chars()).toBeInstanceOf(Iter)
  expect(iter.chars('a')).toBeInstanceOf(Iter)
  expect(iter.chars('a', 'z')).toBeInstanceOf(Iter)

  {
    const it = iter.chars('a', 'z')

    expect(it.producer).toBeInstanceOf(CharProducer)
  }
})

test('.cycle()', () => {
  const it = iter.cycle([1, 2, 3])

  expect(it).toBeInstanceOf(Iter)
  expect(it.producer).toBeInstanceOf(CycleProducer)
})

test('.from()', () => {
  expect(iter.from()).toBeInstanceOf(Iter)
  expect(iter.from([1, 2, 3])).toBeInstanceOf(Iter)
})

test('.of()', () => {
  expect(iter.of()).toBeInstanceOf(Iter)

  {
    const it = iter.of(1, 2, 3)

    expect(it).toBeInstanceOf(Iter)
    expect(it.producer).toMatchSnapshot()
  }
})

test('.range()', () => {
  expect(iter.range()).toBeInstanceOf(Iter)
  expect(iter.range(1)).toBeInstanceOf(Iter)
  expect(iter.range(1, 2)).toBeInstanceOf(Iter)

  {
    const it = iter.range(1, 2)

    expect(it.producer).toBeInstanceOf(NumberProducer)
  }
})

test('.repeat()', () => {
  expect(iter.repeat()).toBeInstanceOf(Iter)
})
