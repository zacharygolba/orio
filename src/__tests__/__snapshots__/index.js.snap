// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.of() 1`] = `
IndexedProducer {
  "size": 3,
  "source": Array [
    1,
    2,
    3,
  ],
  "state": 0,
}
`;
